tartufo/scanner.py

```diff
@@ -90,35 +91,269 @@ def file_path(self) -> Optional[str]:
+            output.append(util.style_ok("Detail: {}".format(self.issue_detail)))
         if self.file_path:
-            output.append(style_ok("Filepath: {}".format(self.file_path)))
+            output.append(util.style_ok("Filepath: {}".format(self.file_path)))
         if self.signature:
-            output.append(style_ok("Signature: {}".format(self.signature)))
+            output.append(util.style_ok("Signature: {}".format(self.signature)))
         if self.branch_name:
-            output.append(style_ok("Branch: {}".format(self.branch_name)))
+            output.append(util.style_ok("Branch: {}".format(self.branch_name)))
         if self.commit:
-            output.append(style_ok("Date: {}".format(self.commit_time)))
-            output.append(style_ok("Hash: {}".format(self.commit_hash)))
-            output.append(style_ok("Commit: {}".format(self.commit_message)))
+            output.append(util.style_ok("Date: {}".format(self.commit_time)))
+            output.append(util.style_ok("Hash: {}".format(self.commit_hash)))
+            output.append(util.style_ok("Commit: {}".format(self.commit_message)))
 
         output.append(diff_body)
         output.append(self.OUTPUT_SEPARATOR)
         return "\n".join(output)
 
 
+class ScannerBase(abc.ABC):
+    _issues: Optional[List[Issue]] = None
+    _included_paths: Optional[List[Pattern]] = None
+    _excluded_paths: Optional[List[Pattern]] = None
+    _rules_regexes: Optional[Dict[str, Pattern]] = None
+    options: GlobalOptions
+
+    def __init__(self, options: GlobalOptions) -> None:
+        self.options = options
+
+    @property
+    def issues(self) -> List[Issue]:
+        if self._issues is None:
+            self._issues = self.scan()
+        return self._issues
+
+    @property
+    def included_paths(self) -> List[Pattern]:
+        if self._included_paths is None:
+            if self.options.include_paths:
+                self._included_paths = config.compile_path_rules(
+                    self.options.include_paths.readlines()
+                )
+            else:
+                self._included_paths = []
+        return self._included_paths
+
+    @property
+    def excluded_paths(self) -> List[Pattern]:
+        if self._excluded_paths is None:
+            if self.options.exclude_paths:
+                self._excluded_paths = config.compile_path_rules(
+                    self.options.exclude_paths.readlines()
+                )
+            else:
+                self._excluded_paths = []
+        return self._excluded_paths
+
+    @property
+    def rules_regexes(self):
+        if self._rules_regexes is None:
+            self._rules_regexes = config.configure_regexes(
+                self.options.default_regexes,
+                self.options.rules,
+                self.options.git_rules_repo,
+                self.options.git_rules_files,
+            )
+        return self._rules_regexes
+
+    @lru_cache()
```

Comment:
I don't have any metrics to back it up, but I'm pretty sure it will.

Imagine a repository with thousands of commits, and only 10 files. This function will get called with the same set of arguments hundreds or even thousands of times, and have to re-check the entire list of regexes every time. This should be a nice shortcut for that.